### ISSUE TITLE
Not support MV3 native ES background worker

### ISSUE THREAD

**Issue by Jack-Works** [awesome-webextension/webpack-target-webextension#24]
Blocked by https://github.com/w3c/ServiceWorker/issues/1356, otherwise, we need to emit import statements which is not easy in a webpack bundle.

**Comment by fregante** [comment]
> I think that maybe it already works with this, regardless if manifest version:

Yep: https://github.com/pixiebrix/pixiebrix-extension/pull/7472

**Comment by Jack-Works** [comment]
I'm unsure if your comment relates to this issue, but I'll write down what I recently found as a memo.

Firstly, Chrome supports service workers with native ES modules now, you need to write your manifest like this:

```json
  "background": {
    "service_worker": "background.worker.js",
    "type": "module"
  },
```

Then the service worker can use `import` or `export` statements, not `importScripts`.

Currently, we use JSONP-style (the default of webpack) to load chunks, this way works well, so even if this issue is not resolved, this plugin is ok to use.

```js
// JSONP-style chunk format
"use strict";
(globalThis["webpackChunk"] = globalThis["webpackChunk"] || []).push(...)
```

As an ES Module enthusiast, I hope we can support ES Module as the chunk format, and also as the chunk loading format.

```js
// ES Module chunk format
export const id = 232;
export const ids = [232];
export const modules = {

/***/ 232:
/***/ (() => {
/***/ })

};
```

If the chunk format becomes ESM, the only way to load it is static `import` or dynamic `import`.

Dynamic `import` already works to load JSONP-style chunks in the content script (https://github.com/awesome-webextension/webpack-target-webextension#content-script), [the problem is dynamic import is not allowed in an ES Module service worker](https://github.com/w3c/ServiceWorker/issues/1356) _even if_ the dynamic import is called before first event loop ends. This requires me to emit not dynamic `import`, but static imports. Webpack now only supports dynamic import (see `__webpack_require__.f.j` if you're interested) those chunks.

This becomes harder, but not impossible. I think it needs a lot of time to figure out how to make this work. I can emit import statements at the top level, but I need to know their _file name_, but unluck, file names are generated from a runtime function (`__webpack_require__.u`) and it's very hard to do it in the compile time.

Since everything works well today, this is not very urgent to fix.

**Comment by fregante** [comment]
> Then the service worker can use import or export statements, not importScripts.

That's very good to know. Parcel forced me to use that syntax but never investigated it. I'll start using it in webpack too because importScript breaks "Pause on uncaught errors" in the dev tools, because the debugger pauses on importScript, regardless of the actual position and source maps.

**Comment by fregante** [comment]
> dynamic `import`

That didn't seem to work for me 🤷‍♂️ I got the same error as before

<img width="288" alt="Screenshot" src="https://github.com/awesome-webextension/webpack-target-webextension/assets/1402241/641b27ae-e382-427c-b147-a30aae7c6875">

Changing it to `import` statement fixed it.

---

So technically now I'm using a native ES background worker in `webpack-target-webextension`. This issue is fixed? Maybe it just needs some documentation?

**Comment by Jack-Works** [comment]
Native ES service worker works, dynamic import does not.

While using this plugin, compiled ES modules (including dynamic import) work after bundling, loading by `importScritps`. Native ES service worker does not work with this plugin currently (the screenshot you gave).

**Comment by Jack-Works** [comment]
I just took another look of this. This is not possible until there is ES Module support in content script.

**Comment by fregante** [comment]
> e is ES Module support in content script

I think you meant "in background workers" right?

**Comment by Jack-Works** [comment]
> > e is ES Module support in content script
> 
> I think you meant "in background workers" right?

No, to share chunks between the background and content script, they must use the format that both environments support. Now ES Module is only supported in the background worker and not supported in the content script, so there is no meaning in investigating this.

**Comment by fregante** [comment]
What do you mean? I definitely am using both static and dynamic ESM in content scripts:

https://robwu.nl/crxviewer/?crx=https%3A%2F%2Fchromewebstore.google.com%2Fdetail%2Frefined-github%2Fhlepfoohegkhhmjieoechaddaejaokhf%3Fhl%3Den

See `content-script.js`.

There are some tricks though:

- you need to use `import(chrome.runtime.getURL('actual-content-script.js'))` in your entry point in order to later be able to use `import` statements
- all JS files need to be in `web_accessible_resources`

**Comment by Jack-Works** [comment]
hmm wait you're right, an extra file will be enough to load ESM in content script.
